### DISCUSSION TITLE
Inspect the feasibility of using a code analyzer to decrease run-time errors

### DISCUSSION THREAD

**Discussion by jonasraoni** [pkp/pkp-lib#7973] [Proposals]
**Describe the problem you would like to solve**
Due to the nature of PHP, it's challenging to detect simple run-time errors due to broken typing promises.

**Describe the solution you'd like**
Investigate whether [Psalm](https://psalm.dev), is able to help us mitigate such errors at building-time.
Perhaps it will produce too much noise, due to untyped 3rd-party code ([they do have plugins to improve the type handling for some libraries](https://psalm.dev/plugins)).

**Who is asking for this feature?**
Developer

**Comment by NateWr** [comment]
Were you thinking of trying to use it during the validation tests with Travis?

**Comment by jonasraoni** [comment]
Yeah, given that we don't have a tons of unit tests, anything that helps mitigating unexpected errors automatically, sounds great for me.
But just like the issue about "removing deprecated code", this is mainly a self-reminder. Let me know if we have a better place for this kind of thing :)

**Comment by NateWr** [comment]
I have a bunch of these note-to-self issues scattered around too. As long as it's something you really plan to get to, I'm ok leaving them filed here.

**Comment by jonasraoni** [comment]
Ok! So I'll use this issue to keep related resources. In order to auto-refactor code to use newer PHP features (it might probably cover deprecations), we can use the https://github.com/rectorphp/rector

**Comment by jonasraoni** [comment]
Here's an article comparing [Psalm](https://github.com/vimeo/psalm), [PHPStan](https://github.com/phpstan/phpstan) and [Phan](https://github.com/phan/phan):
https://medium.com/bumble-tech/php-code-static-analysis-based-on-the-example-of-phpstan-phan-and-psalm-a20654c4011d

Like Typescript, such tools might have plugins to help with the types of popular libraries:
Psalm + Laravel: https://github.com/psalm/psalm-plugin-laravel
PHPStan + Laravel: https://github.com/nunomaduro/larastan
